cmd/bamTagToReadGroup/bamTagToReadGroup.go

```diff
@@ -0,0 +1,141 @@
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/sam"
+	"io"
+	"log"
+	"strings"
+	"sync"
+)
+
+func usage() {
+	fmt.Print(
+		"bamTagToReadGroup - Add bam read groups based on read tags\n\n" +
+			"Usage:\n" +
+			"  bamTagToReadGroup [options] -tagId CB -tagValues barcodes.tsv -i input.bam > output.bam\n\n" +
+			"Options:\n")
+	flag.PrintDefaults()
+}
+
+func main() {
+	var infile *string = flag.String("i", "", "Input bam file.")
+	var outfile *string = flag.String("o", "stdout", "Output bam file.")
+	var tagId *string = flag.String("tagId", "", "Bam tag ID that will be used to form read groups.")
+	var tagValuesFile *string = flag.String("tagValues", "", "File of tag values that will be added to read groups, one value per line. "+
+		"If bam record has a tagId value matching an entry in this file, the record will be annotated as belonging to a read group with the same name as the tag value. "+
+		"If a bam record does not have a tag with tagId, or the tagId value does not match an entry in this file then the bam record will not be assigned a read group."+
+		"NOTE: All existing read groups will be removed from the file.")
+	flag.Parse()
+	flag.Usage = usage
+
+	if *infile == "" {
+		usage()
+		log.Fatalln("ERROR: must provide a bam file with -i")
+	}
+
+	if *tagId == "" || len(*tagId) != 2 {
+		usage()
+		log.Fatalln("ERROR: -tagId must be 2 characters")
+	}
+
+	if *tagValuesFile == "" {
+		usage()
+		log.Fatalln("ERROR: must provide a tag values file with -tagValues")
+	}
+
+	bamTagToReadGroup(*infile, *outfile, *tagId, *tagValuesFile)
+}
+
+func bamTagToReadGroup(infile, outfile, tagId, tagValuesFile string) {
+	tagValues := fileio.Read(tagValuesFile)
+	tagValueMap := make(map[string]bool)
+	for i := range tagValues {
+		tagValueMap[tagValues[i]] = true
+	}
+
+	inputRecordChan, header := sam.GoReadToChan(infile)
+	addTagsToHeader(&header, tagValues)
+
+	out := fileio.EasyCreate(outfile)
+	defer safeclose(out)
+	bw := sam.NewBamWriter(out, header)
+	defer safeclose(bw)
+
+	// spawn goroutine to write records after being updated
+	writeChan := make(chan sam.Sam, 1000)
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go func(writeChan <-chan sam.Sam, bw *sam.BamWriter, wg *sync.WaitGroup) {
+		for rec := range writeChan {
+			sam.WriteToBamFileHandle(bw, rec, 0)
+		}
+		wg.Done()
+	}(writeChan, bw, wg)
+
+	// read records and update read groups as we go
+	for rec := range inputRecordChan {
+		updateRecord(&rec, tagId, tagValueMap)
+		writeChan <- rec
+	}
+	close(writeChan)
+
+	// wait for goroutine to finish writing before return on main thread
+	wg.Wait()
+}
+
+func safeclose(f io.Closer) {
+	err := f.Close()
+	exception.PanicOnErr(err)
+}
+
+// addTagsToHeader adds each element in tagValues as a read group in the header
+// format is `@RG	ID:$TAGVALUE	SM:$TAGVALUE	LB:$TAGVALUE`
+func addTagsToHeader(header *sam.Header, tagValues []string) {
```

Comment:
Seems a bit cumbersome, but I think using string builder to concat strings would in theory is more performant.

```go
import(
    "strings"
)

func addTagsToHeader(header *sam.Header, tagValues []string) {
    var str strings.Builder

    // Iterate over existing header lines
    for _, line := range header.Text() {
        if !strings.HasPrefix(line, "@RG") {
            str.WriteString(line)
            str.WriteString("\n")
        }
    }

    // Add new @RG lines
    for _, value := range tagValues {
        str.WriteString("@RG\tID:")
        str.WriteString(value)
        str.WriteString("\tSM:")
        str.WriteString(value)
        str.WriteString("\tLB:")
        str.WriteString(value)
        str.WriteString("\n")
    }

    // Update the header text
    header.SetText(str.String())
}
```

Reply:
I didn't perform an checks or test with this suggested block so be sure to check if it's right

Reply:
It would be more performant, but the computational expense for processing the header is essentially negligible compared to processing of the bam records. In this case, I think it makes sense to keep the simpler and slightly worse implementation IMO.